src/core/public/application/application_service.tsx

```diff
@@ -304,10 +311,16 @@ export class ApplicationService {
         return absolute ? relativeToAbsolute(relUrl) : relUrl;
       },
       navigateToApp,
-      navigateToUrl: async (url) => {
+      navigateToUrl: async (
+        url: string,
+        { skipAppLeave = false, forceRedirect = false }: NavigateToUrlOptions = {}
+      ) => {
         const appInfo = parseAppUrl(url, http.basePath, this.apps);
-        if (appInfo) {
-          return navigateToApp(appInfo.app, { path: appInfo.path });
+        if (forceRedirect === true) {
+          window.removeEventListener('beforeunload', this.onBeforeUnload);
+          return this.redirectTo!(url);
+        } else if (appInfo) {
+          return navigateToApp(appInfo.app, { path: appInfo.path, skipAppLeave });
         } else {
           return this.redirectTo!(url);
         }
```

Comment:
Hum, I think the condition for `window.removeEventListener` is wrong. We should be checking for `skipAppLeave` here, not `forceRedirect`:

```ts
if (forceRedirect || !appInfo) {
    if(skipAppLeave) {
        window.removeEventListener('beforeunload', this.onBeforeUnload);
    }
    return this.redirectTo!(url);
}
return navigateToApp(appInfo.app, { path: appInfo.path, skipAppLeave });
```

Reply:
great chat!